src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/TempPE/TempPEBuildManager.cs

```diff
@@ -0,0 +1,550 @@
+                    await FireTempPEDirtyAsync(item, value.ShouldCompile);
+                }
+            }
+
+            // Shared items cause all TempPEs to be dirty, but don't recompile, to match legacy behaviour
+            if (hasRemovedDesignTimeSharedInputs || hasAddedDesignTimeSharedInputs)
+            {
+                // adding or removing shared design time inputs dirties things but doesn't recompile
+                foreach (string item in newDesignTimeInputs)
+                {
+                    // We don't want to fire again if we already fired above and compiled
+                    if (!addedDesignTimeInputs.Contains(item))
+                    {
+                        await FireTempPEDirtyAsync(item, false);
+                    }
+                }
+            }
+
+            foreach (string item in removedDesignTimeInputs)
+            {
+                _buildManager.Value.OnDesignTimeOutputDeleted(item);
+            }
+
+            bool TryGetValueIfUnused<T>(string item, ImmutableDictionary<string, T>.Builder source, ImmutableHashSet<string>.Builder otherSources, out T result)
+            {
+                // return the value from the source, but only if it doesn't appear in other sources
+                return source.TryGetValue(item, out result) && !otherSources.Contains(item);
+            }
+
+        }
+
+        protected virtual ITaskDelayScheduler CreateTaskScheduler()
+        {
+            return new TaskDelayScheduler(s_compilationWaitTime, UnconfiguredProjectServices.ThreadingService, DisposalToken);
+        }
+
+        protected override async Task DisposeCoreAsync(bool initialized)
+        {
+            if (initialized)
+            {
+                DisposeTaskSchedulers();
+            }
+            await base.DisposeCoreAsync(initialized);
+        }
+
+        /// <summary>
+        /// InitializeInnerCoreAsync is responsible for setting an initial AppliedValue. This value will be used by any UI thread calls that may happen
+        /// before the first data flow blocks have been processed. If this method doesn't return a value then the system will block until the first blocks
+        /// have been applied but since we are initialized on the UI thread, that blocks for us, so we must return an initial state.
+        /// </summary>
+        protected override Task InitializeInnerCoreAsync(CancellationToken cancellationToken)
+        {
+            AppliedValue = new ProjectVersionedValue<DesignTimeInputsItem>(new DesignTimeInputsItem(), ImmutableDictionary.Create<NamedIdentity, IComparable>());
+
+            return Task.CompletedTask;
+        }
+
+        /// <summary>
+        /// This method is where we tell data flow which blocks we're interested in receiving updates for
+        /// </summary>
+        protected override IDisposable LinkExternalInput(ITargetBlock<IProjectVersionedValue<InputTuple>> targetBlock)
+        {
+            return ProjectDataSources.SyncLinkTo(
```

Comment:
I think you'll need to JoinUpstreamDataSources on SourceItemsRuleSource and ProjectRuleSource, so that if someone blocks on the UI thread for your result and these need to finish on the UI thread to complete, that be allowed to proceed because there's a relationship between yourself and them.